src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -343,26 +395,58 @@ private bool IsRecovering(ServerErrorCode code, string message)
             return false;
         }
 
+        private bool IsRecoveringErrorException(Exception exception)
+        {
+            return
+                exception is MongoCommandException commandException &&
+                IsRecovering((ServerErrorCode)commandException.Code, commandException.ErrorMessage);
+        }
+
+        private void SetDescription(ServerDescription newDescription)
+        {
+            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, null, null);
+            SetDescription(oldDescription, newDescription);
+        }
```

Comment:
Alternatively, I'm not sure, but I feel like this may be slightly less race prone since the compare and exchange happens as an atomic operation?
```
        private void SetDescription(ServerDescription newDescription)
        {
            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, newDescription, _currentDescription);
            OnDescriptionChanged(sender: this, new ServerDescriptionChangedEventArgs(oldDescription, newDescription));
        }
```
What do you guys think?

Reply:
After thinking this carefully over the weekend, I do believe that the version proposed in the above comment is less race prone. As originally implemented, the following is possible: 

Thread 1= T1, Thread2= T2.  State = `_currentDescription`.
1. T1: State: A, New: B, read State into `oldDescription` and create (A->B) ServerDescriptionChangedEventArgs (without raising it)
2. T2: State: A, New: C, read State into `oldDescription` (A), create (A->C) ServerDescriptionChangedEventArgs
3. T2: Write `newDescription` (C) into State (which contains A). State transitions from A to C.
4. T2: Raise (A->C) event
5. T1: Write B into state (which currently contains C). State goes from  C to B
6. T1: Raise (A->B) event **(this transition never actually happened)**

Now, using the new version: 

T1: New: B. State: A. Atomically set state to B and store A into `old`.  
T2: New: C. State: B. Atomically set state to C and store B to `old`. 
T2: Create a B->C event using `old: B` and `new: C`, raise event that a (B->C) change happened
T1: Ditto above but with `old: A` and `new: B`. Raise event that (A->B) happened.

The events DID happen (but were raised out of order) which is an improvement 
from raising events that did not happen
 
The unit tests pass and EG is green with this change so I will include this change in my next set of commits and link back to this comment.